### PR TITLE
fix(amazonq): Emit telemetry events when operations fail for /dev feature

### PR DESCRIPTION
**Pull request by sarayev**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Description

So far, telemetry events are only emitted when operations succeed for /dev feature, with this PR we're emitting telemetry events even when the operations fail. 

We're also adding result, reason, duration, credentialStartUrl metadata fields to the emitted events. 

## Checklist
- [x] My code follows the code style of this project
- [N/A] I have added tests to cover my changes - Adjusted existing tests
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE. - I think it is not customer facing change, thus not added Changlog entry. 
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/CodeGenerationState.kt [hunk 1]
```diff
@@ -27,49 +33,65 @@ class CodeGenerationState(
...
+                nextState = nextState,
+                interaction = Interaction(content = "", interactionSucceeded = true)
+            )
+        } catch (e: Exception) {
+            logger.warn(e) { "$FEATURE_NAME: Code generation failed: ${e.message}" }
+            result = Result.Failed
+            failureReason = e.message
+            throw e
+        } finally {
+            AmazonqTelemetry.codeGenerationInvoke(
```
Comment (kumsmrit) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/CodeGenerationState.kt:80]: Do we need to add failure telemetry log here as well, or just adding in API call failures (`FeatureDevClientUtil`) is enough ?

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/CodeGenerationState.kt [hunk 2]
```diff
@@ -27,49 +33,65 @@ class CodeGenerationState(
...
+                amazonqConversationId = config.conversationId,
+                amazonqCodeGenerationResult = CodeGenerationWorkflowStatus.COMPLETE.toString(),
```
Comment (sarayev) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/CodeGenerationState.kt:82]: note to myself - this has to be changed.

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/CodeGenerationState.kt (hunk 1 above)
Comment (sarayev) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/CodeGenerationState.kt:80]: I think in this event, we're interested in how long the code gen took and if the codegen failed or not. Not only the invocation of startTaskAssistCodeGeneration API. See the implementation in vscode: https://github.com/aws/aws-toolkit-vscode/blob/master/packages/core/src/amazonqFeatureDev/session/sessionState.ts#L304

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/util/FeatureDevClientUtil.kt [hunk 3]
```diff
@@ -28,189 +29,215 @@ import software.aws.toolkits.jetbrains.services.amazonqFeatureDev.clients.Genera
...
-        logger.warn(e) { "$FEATURE_NAME: Failed to generate presigned url: ${e.message}" }
 
-        var errMssg = e.message
-        if (e is CodeWhispererRuntimeException) {
-            errMssg = e.awsErrorDetails().errorMessage()
-            logger.warn(e) { "Create UploadUrl failed for request: ${e.requestId()}" }
-
-            if (e is ValidationException && e.message?.contains("Invalid contentLength") == true) {
-                throw ContentLengthError(message("amazonqFeatureDev.content_length.error_text"), e.cause)
+    fun createUploadUrl(conversationId: String, contentChecksumSha256: String, contentLength: Long):
```
Comment (kumsmrit) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/util/FeatureDevClientUtil.kt:82]: Do we not need to add telemetry for `createUploadUrl` API ?

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/SessionState.kt [hunk 4]
```diff
@@ -3,10 +3,13 @@
 
 package software.aws.toolkits.jetbrains.services.amazonqFeatureDev.session
 
+import software.aws.toolkits.jetbrains.services.amazonqFeatureDev.util.FeatureDevClientUtil
+
 interface SessionState {
     val tabID: String
     val phase: SessionStatePhase?
     var approach: String
+    val featureDevClientUtil: FeatureDevClientUtil
```
Comment (kumsmrit) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/SessionState.kt:12]: Should we define this `featureDevClientUtil` in SessionStateConfig instead; attaching this with `SessionState` here makes it a bit misleading.

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/util/FeatureDevService.kt [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+                }
+            }
+            apiError(errMssg, e.cause)
+        }
+    }
+
+    suspend fun generatePlan(
+        conversationId: String,
+        uploadId:
+        String,
```
Comment (kumsmrit) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/util/FeatureDevService.kt:114]: nit: Type can be with the argument definition above `uploadId: String`

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/util/FeatureDevService.kt [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+            logger.debug { "Executing generateTaskAssistPlan with conversationId $conversationId" }
+            val generatePlanResult = proxyClient.generateTaskAssistPlan(
+                conversationId,
+                uploadId,
+                message
+            )
+            return generatePlanResult
+        } catch (e: Exception) {
+            logger.warn(e) { "$FEATURE_NAME: Failed to execute planning : ${e.message}" }
+            failureReason = e.javaClass.simpleName
```
Comment (kumsmrit) [inline plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/util/FeatureDevService.kt:131]: Should this `failureReason` be updated based on specific exception handling for ServiceQuotaExceededException ?

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/util/FeatureDevClientUtil.kt (hunk 3 above)
Comment (sarayev) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/util/FeatureDevClientUtil.kt]: It is added in a different place, right? And it captures duration/errors for the zipping part too.

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/SessionState.kt (hunk 4 above)
Comment (sarayev) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/SessionState.kt]: Yeah, as discussed offline, renamed it to be FeatureDevService and moved the parameter to the config.

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/util/FeatureDevService.kt (hunk 6 above)
Comment (kumsmrit) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/util/FeatureDevService.kt:130]: I think for VSCode we explicitly show `PlanIterationLimitError`; but we could keep it as is here given the `ServiceQuotaExceededException` will be specific for different APIs like createConversation, planning and codegen

**Review by kumsmrit — APPROVED** [review]

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/util/FeatureDevService.kt [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+            return uploadUrlResponse
+        } catch (e: Exception) {
+            logger.warn(e) { "$FEATURE_NAME: Failed to generate presigned url: ${e.message}" }
+
+            var errMssg = e.message
+            if (e is CodeWhispererRuntimeException) {
+                errMssg = e.awsErrorDetails().errorMessage()
+                logger.warn(e) { "Create UploadUrl failed for request: ${e.requestId()}" }
+
+                if (e is ValidationException && e.message?.contains("Invalid contentLength") == true) {
```
[the remark this answers is not in the document]
Comment (sarayev) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/util/FeatureDevService.kt:103]: We already have a test for it. `test createUploadUrl with validation error` test in FeatureDevServiceTest.kt.

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/util/FeatureDevService.kt [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+            return startCodeGenerationResponse
+        } catch (e: Exception) {
+            logger.warn(e) { "$FEATURE_NAME: Failed to execute startTaskAssistCodeGeneration ${e.message}" }
+
+            var errMssg = e.message
+            if (e is CodeWhispererRuntimeException) {
+                errMssg = e.awsErrorDetails().errorMessage()
+                logger.warn(e) { "StartTaskAssistCodeGeneration failed for request: ${e.requestId()}" }
+
+                if (e is software.amazon.awssdk.services.codewhispererruntime.model.ServiceQuotaExceededException || (
```
[the remark this answers is not in the document]
Comment (sarayev) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/util/FeatureDevService.kt:178]: We need the entire package name, unfortunately. We already import `import software.amazon.awssdk.services.codewhispererstreaming.model.ServiceQuotaExceededException` with the same name and use it above. So, we can't have 2 imports with the same name.

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/CodeGenerationState.kt [hunk 9]
```diff
@@ -27,49 +31,68 @@ class CodeGenerationState(
...
+                amazonqCodeGenerationResult = codeGenerationWorkflowStatus.toString(),
+                amazonqGenerateCodeIteration = currentIteration.toDouble(),
+                amazonqNumberOfReferences = numberOfReferencesGenerated?.toDouble(),
+                amazonqGenerateCodeResponseLatency = (System.currentTimeMillis() - startTime).toDouble(),
+                amazonqNumberOfFilesGenerated = numberOfFilesGenerated?.toDouble(),
+                amazonqRepositorySize = repositorySize,
+                result = result,
+                reason = failureReason,
+                duration = (System.currentTimeMillis() - startTime).toDouble(),
+                requestServiceType = "amazoncodewhispererservice",
```
[the remark this answers is not in the document]
Comment (sarayev) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/session/CodeGenerationState.kt]: I removed `requestServiceType` field from these PR. I don't see it being used by any of our dashboards atm. Can discuss it with the team, and add it separately if we need it.

**Review by manodnyab — APPROVED** [review]

File: plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/util/FeatureDevService.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (manodnyab) [reply plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/util/FeatureDevService.kt:176]: You can probably use an alias here

**Pull request merged**